File: Week_05/Practicum/Sliding Window Maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> q;
        vector<int> res;

        for (int i = 0; i < k; i++) {
            while(!q.empty() && nums[q.back()] <= nums[i]) {
                q.pop_back();
            }
            q.push_back(i);
        }

        res.push_back(nums[q.front()]);

        for (int i = k; i < nums.size(); i++) {
            if (q.front() <= i - k) {
                q.pop_front();
            }

            while(!q.empty() && nums[q.back()] <= nums[i]) {
                q.pop_back();
            }
            q.push_back(i);

            res.push_back(nums[q.front()]);
        }
        return res;
    }
};

File: Week_05/Practicum/python/Sliding Window Maximum.py
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        ans = []
        dq = deque()

        for i in range(len(nums)):
            # Remove all numbers in the back, which are less than the incoming
            # We keep it decreasing that way, so that the front is always the answer
            while dq and dq[-1] < nums[i]:
                dq.pop()
                
            dq.append(nums[i])

            # We are removing the max
            if i >= k and dq[0] == nums[i - k]:
                dq.popleft()
            
            if i >= k - 1:
                ans.append(dq[0])

        return ans

